iMX RT full pinctrl enablement

danieldegrasse opened the pull request:
This PR enables the pin control API for all in tree iMX RT boards and drivers. It includes the following changes:
- driver level pinctrl enablement for all drivers used by iMX RT SOCs
- new pinctrl driver for RT11xx series, based on the RT10xx series pinctrl driver
- updates to the RT10xx series pinctrl driver to support GPR register, which is required for some pinmux settings
- updates to iMX RT GPIO driver to enable pinmux selection, by leveraging pinctrl driver nodes

~~Dependent upon #43293 for missing GPIO definitions in RT1170 devicetree~~

danieldegrasse committed:
boards: add pinctrl definitions for iMX RT boards

Add pinctrl definitions for all iMX RT based boards

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: pinctrl: Enable pinctrl for RT11xx series

RT11xx series has similar pin configuration peripheral to RT10xx, with
some differences in register layout. Create new pinctrl definition
header file, and reuse existing driver code for RT10xx.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: mcux-12b1-adc: enable pinctrl for mcux-12b1 adc

Enable pinctrl for mcux_12b1msps_sar adc driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: lpc_lpadc: enable pinctrl for lpadc driver

enable pinctrl for mcux lpc lpadc driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: pwm: enable pinctrl for nxp imx_pwm driver

Enable pinctrl for nxp imx_pwm driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: memc: enable pinctrl for flexspi driver

Enable pinctrl for flexspi driver. Note that when flexspi is being using
in XIP mode, pinctrl settings are not required and will not be applied.
Pinctrl settings are only required when the flexspi device being used is
not the one used for XIP.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: video: video_mcux_csi: enable pinctrl

Enable pinctrl for mcux_csi video driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: display: display_mcux_elcdif: add pinctrl support

add pinctrl support for display_mcux_elcdif driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: disk: usdhc: enable pinctrl for usdhc driver

Enable pinctrl driver for usdhc. USDHC driver uses custom pinctrl states
for fast, slow, and medium signal frequencies, as well as pin pull for
SD detection.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: pinctrl: update rt pinctrl to support GPR register

iMX.RT parts use a GPR register for some pinmux settings. Update pinctrl
driver to support this GPR register definition.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: gpio: gpio_mcux_igpio: enable pinmuxing in gpio driver

In order for pinctrl support to be complete, RT series GPIO driver must
support pinmuxing within the driver level. RT series pinmux settings do
not correspond directly to gpio port/pin numbers, so use DTS mappings to
pinctrl nodes to select and apply pinmux settings in the gpio driver.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: display: display_mcux_elcdif: add backlight-gpios property

Add backlight gpios property to mcux display driver, so that the driver
can correctly initialize the backlight gpio control.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: ethernet: eth_mcux: add support for resetting phy

add support for resetting phy at boot via GPIO reset pin.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: nxp_imx: rt: enable ethernet ref clock output

Enable soc ethernet ref clock output at the SOC level instead of board
level, since it is required for all iMX.RT SOCs

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: imxrt: remove pinmux configuration for all imxrt boards

imx.rt boards support pinctrl. Remove pinmux settings

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
manifest: update NXP HAL in west.yml

Update nxp hal revision in west.yml to include new pinctrl definition
header files.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
@hakehuang Can you run CI against this PR? It should only need to be tested on iMX RT boards

hakehuang commented:
> @hakehuang Can you run CI against this PR? It should only need to be tested on iMX RT boards

CI is running, will feedback once finish

gmarull commented on drivers/disk/usdhc.c:
```diff
@@ -1910,6 +1922,31 @@ static void usdhc_enable_ddr_mode(USDHC_Type *base,
+	bool init, uint32_t speed, uint32_t strength)
+{
+#ifdef CONFIG_PINCTRL
+	/*
```
nit: doesn't this branch require `ARG_UNUSED()` for some of the parameters? I find this function a bit confusing for the pinctrl case, callers are e.g. passing `CARD_BUS_STRENGTH_7` but it is in fact ignored. Maybe creating a separate helper would be better.

danieldegrasse replied:
A separate helper would work fine as well. The reason I used this approach was to limit the locations a conditional `#ifdef CONFIG_PINCTRL` was used in the code. There's only a few calls to this helper, so I'll switch the implementation to use `pinctrl_apply_state` directly there.

The function only truly needs the frequency to decide what pin settings to select (the SDK always uses a drive strength of 7 for USDHC pins), but the NXP SDK API used to configure pins accepts a drive strength, so this parameter was made available to the user.

gmarull commented on dts/bindings/display/nxp,imx-elcdif.yaml:
```diff
@@ -102,3 +102,9 @@ properties:
       description:
         LCD data bus width. The default is set to the reset value of 16-bit
+
+    backlight-gpios:
```
nit: you can also use the regulator framework so that devices don't need any modification (if never controlled).

danieldegrasse replied:
It is in fact driving the backlight that causes issues- see here: https://github.com/zephyrproject-rtos/zephyr/issues/26527#issuecomment-678524116. We cannot use the regulator framework for this.

gmarull replied:
np, was just a suggestion

danieldegrasse replied:
No worries, just glad I checked- I wasn't aware of the full extend of the backlight issue, and the regulator approach would have been cleaner if the backlight problem wasn't present.

gmarull reviewed (commented):
Overall LGTM, I have a couple of minor questions

danieldegrasse committed:
soc: nxp_imx: rt: remove unnecessary pinmux callbacks

All iMX RT 1xxx boards are now expected to use pin control, so the USDHC
pinmuxing callbacks can be removed.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

MaureenHelm reviewed (dismissed):
Nice work!
